File: breath_first_search/breath_first.py
from collections import deque

search = deque()

graph = {}

graph["you"] = ["alice", "bob", "victor"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "monny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []


def person_is_seller(name):
    return name[-1] == 'm'

def search(name):
    search_queue = deque()
    search_queue += graph[name]
    searched = []
    while search_queue:
        person = search_queue.popleft()

        if not person_is_seller(person):
            print(f"{person} is not a mango seller")
            search_queue += graph[person]
            searched.append(person)
        else:
            print(f"{person} is a mango seller!")
            return True
    print("No mango sellers found!")
    return False

search("you")

File: breath_first_search/breath_first.js
const graph = {
  you: ["alice", "bob", "victor"],
  bob: ["anuj", "peggy"],
  alice: ["peggy"],
  claire: ["thom", "monny"],
  anuj: [],
  peggy: [],
  thom: [],
  jonny: [],
};

function person_is_seller(name) {
  return name.charAt(name.length - 1) === "m";
}

function search(name) {
  const searchQueue = [];
  searchQueue.push(...graph[name]);
  const searched = [];

  while (searchQueue.length) {
    const person = searchQueue.shift();

    if (searched.includes(person)) {
      continue;
    }

    if (person_is_seller(person)) {
      console.log(`${person} is a mango seller!`);
      return true;
    } else {
      console.log(`${person} is not a mango seller`);
      const connections = graph[person];
      if (Array.isArray(connections)) {
        searchQueue.push(...connections);
      }
      searched.push(person);
    }
  }

  console.log("No mango sellers found!");
  return false;
}

search("you");
